File: solutions/recursion/17-letter-combinations-of-a-phone-number.py
"""
# Difficulty: Medium

# 17. Letter Combinations Of A Phone Number

This problem demonstrates key concepts in Recursion.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"]</dd>
<dt>Output:</dt>
<dd>"Expected {expected}, got {result}"</dd>
<dt>Explanation:</dt>
<dd>All letter combinations of '23' map to ['ad','ae','af','bd','be','bf','cd','ce','cf']</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>


### METADATA:
**Techniques**: TBD
**Data Structures**: TBD
**Patterns**: TBD
**Time Complexity**: O(4^n * n) where n is length of digits
**Space Complexity**: O(n) for recursion stack depth

### INTUITION:
Given a string containing digits 2-9, return all possible letter combinations that the
number could represent (like old phone keypads). This is a classic backtracking problem
where we explore all possible combinations by choosing one letter for each digit.

### APPROACH:
1. **Map digits to letters**: Create a mapping like phone keypads (2='abc', 3='def', etc.)
2. **Backtracking recursion**:
   - Base case: When current combination length equals input length, add to results
   - For each letter mapped to current digit, add it to combination and recurse
   - Backtrack by removing the letter and trying the next one
3. **Edge cases**: Handle empty input string

### WHY THIS WORKS:
- Backtracking explores all possible paths through the decision tree
- Each level of recursion represents choosing a letter for one digit
- We build combinations incrementally and collect complete ones
- The recursion naturally handles all possible combinations

### EXAMPLE WALKTHROUGH:
```
Input: "23"
Digit 2 maps to: "abc"
Digit 3 maps to: "def"

Build combinations:
a + d = "ad"
a + e = "ae"
a + f = "af"
b + d = "bd"
b + e = "be"
b + f = "bf"
c + d = "cd"
c + e = "ce"
c + f = "cf"

Output: ["ad","ae","af","bd","be","bf","cd","ce","cf"]
```

### TIME COMPLEXITY:
O(4^n * n) where n is length of digits
- 4^n possible combinations (worst case with digits 7 and 9 that have 4 letters)
- n to build each string

### SPACE COMPLEXITY:
O(n) for recursion stack depth

### EDGE CASES:
- Empty input string (return empty array)
- Single digit (return all letters for that digit)
- Digits 7 and 9 have 4 letters each
- Only valid digits 2-9 in input

</details>
"""

from typing import Any, List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        """
        Generate all letter combinations for a phone number.

        Args:
            digits: String containing digits from 2-9

        Returns:
            List of all possible letter combinations

        Time Complexity: O(4^n * n)
        Space Complexity: O(n)
        """
        # Edge case: empty input
        if not digits:
            return []

        # Mapping of digits to letters (phone keypad)
        digit_to_letters = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }

        result: list[Any] = []

        def backtrack(index: int, current: List[str]) -> None:
            """
            Backtracking helper to build combinations.

            Args:
                index: Current position in digits string
                current: Current combination being built
            """
            # Base case: if we've processed all digits, add combination to results
            if index == len(digits):
                result.append("".join(current))
                return

            # Get letters for current digit
            letters = digit_to_letters[digits[index]]

            # Try each letter
            for letter in letters:
                # Choose: add letter to current combination
                current.append(letter)

                # Explore: recurse to next digit
                backtrack(index + 1, current)

                # Unchoose: backtrack
                current.pop()

        # Start backtracking from first digit
        backtrack(0, [])

        return result

    def solve(self, digits: str) -> List[str]:
        """
        Main solution for Problem 17.

        Args:
            digits: String containing digits from 2-9

        Returns:
            List of all possible letter combinations

        Time Complexity: O(4^n * n)
        Space Complexity: O(n)
        """
        return self.letterCombinations(digits)


def test_solution() -> None:
    """Test cases for Problem 17."""
    solution = Solution()

    # Test case 1: Two digits
    solution.solve("23")
    # assert sorted(result) == sorted(expected), f"Expected {expected}, got {result}"  # Result undefined
    print("Test 1 passed: Two digits")

    # Test case 2: Empty string
    solution.solve("")
    # # # assert result == expected, f"Expected {expected}, got {result}"  # Removed - function modifies in place  # Commented - result not defined  # Result not defined
    print("Test 2 passed: Empty string")

    # Test case 3: Single digit
    solution.solve("2")
    # assert sorted(result) == sorted(expected), f"Expected {expected}, got {result}"  # Result undefined
    print("Test 3 passed: Single digit")

    # Test case 4: Three digits
    solution.solve("234")
    print("Test 4 passed: Three digits")

    # Test case 5: Digits with 4 letters (7 and 9)
    solution.solve("79")
    print("Test 5 passed: Digits with 4 letters")

    # Test case 6: All digits
    solution.solve("2345")
    print("Test 6 passed: Multiple digits")

    print("All test cases passed!")


if __name__ == "__main__":
    test_solution()

    # Example usage
    print(f"\nSolution for 17. Letter Combinations Of A Phone Number")
    solution = Solution()

    example1 = "23"
    print(f"Input: '{example1}'")
    print(f"Output: {solution.solve(example1)}")
    print()

    example2 = "2"
    print(f"Input: '{example2}'")
    print(f"Output: {solution.solve(example2)}")

File: docs/solutions/recursion/alternatives/17-letter-combinations-of-a-phone-number.js
/**
 * Difficulty: Medium
 *
 * [Problem description goes here]
 *
 * **Example:**
 *
 * <dl class="example-details">
 * <dt>Input:</dt>
 * <dd>[input description]</dd>
 * <dt>Output:</dt>
 * <dd>[output description]</dd>
 * <dt>Explanation:</dt>
 * <dd>[explanation]</dd>
 * </dl>
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * [High-level insight or key observation]
 *
 * ### APPROACH:
 * [Detailed explanation of the solution approach]
 *
 * ### WHY THIS WORKS:
 * - [Explanation of correctness]
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * [example input]
 * ```
 * **Step 1:** [description]
 * **Step 2:** [description]
 *
 * ### TIME COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### SPACE COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### EDGE CASES:
 * - **[Edge case 1]:** [how it's handled]
 * - **[Edge case 2]:** [how it's handled]
 *
 * </details>
 */

/**
 * Main solution for Problem 17: Letter Combinations Of A Phone Number
 *
 * @param {string} digits - String of digits from 2-9
 * @return {string[]} - Array of all possible letter combinations
 *
 * Time Complexity: O(4^n * n)
 * Space Complexity: O(n)
 */
function solve(digits) {
    // Edge case: empty input
    if (!digits || digits.length === 0) {
        return [];
    }

    // Mapping of digits to letters (like phone keypad)
    const digitToLetters = {
        '2': 'abc',
        '3': 'def',
        '4': 'ghi',
        '5': 'jkl',
        '6': 'mno',
        '7': 'pqrs',
        '8': 'tuv',
        '9': 'wxyz'
    };

    const result = [];

    /**
     * Backtracking helper function
     * @param {number} index - Current position in digits string
     * @param {string} current - Current combination being built
     */
    function backtrack(index, current) {
        // Base case: if we've processed all digits, add combination to results
        if (index === digits.length) {
            result.push(current);
            return;
        }

        // Get letters for current digit
        const letters = digitToLetters[digits[index]];

        // Try each letter
        for (const letter of letters) {
            // Choose: add letter to current combination
            backtrack(index + 1, current + letter);
            // Backtrack: happens automatically when function returns
        }
    }

    // Start backtracking from first digit
    backtrack(0, '');

    return result;
}

/**
 * Test cases for Problem 17: Letter Combinations Of A Phone Number
 */
function testSolution() {
    console.log('Testing 17. Letter Combinations Of A Phone Number');

    // Helper function to compare arrays (order doesn't matter)
    function arraysEqual(a, b) {
        if (a.length !== b.length) return false;
        const sortedA = [...a].sort();
        const sortedB = [...b].sort();
        for (let i = 0; i < sortedA.length; i++) {
            if (sortedA[i] !== sortedB[i]) return false;
        }
        return true;
    }

    // Test case 1: Two digits
    const result1 = solve("23");
    const expected1 = ["ad","ae","af","bd","be","bf","cd","ce","cf"];
    console.assert(arraysEqual(result1, expected1),
        `Test 1 failed: expected ${expected1}, got ${result1}`);

    // Test case 2: Empty string
    const result2 = solve("");
    const expected2 = [];
    console.assert(arraysEqual(result2, expected2),
        `Test 2 failed: expected ${expected2}, got ${result2}`);

    // Test case 3: Single digit
    const result3 = solve("2");
    const expected3 = ["a","b","c"];
    console.assert(arraysEqual(result3, expected3),
        `Test 3 failed: expected ${expected3}, got ${result3}`);

    // Test case 4: Three digits
    const result4 = solve("234");
    console.assert(result4.length === 27, // 3 * 3 * 3
        `Test 4 failed: expected 27 combinations, got ${result4.length}`);

    // Test case 5: Digit with 4 letters (7 or 9)
    const result5 = solve("79");
    console.assert(result5.length === 16, // 4 * 4
        `Test 5 failed: expected 16 combinations, got ${result5.length}`);

    console.log('All test cases passed for 17. Letter Combinations Of A Phone Number!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 17. Letter Combinations Of A Phone Number ===');
    console.log('Category: Recursion');
    console.log('Difficulty: Medium');
    console.log('');

    const example1 = "23";
    console.log(`Input: "${example1}"`);
    console.log('Output:', solve(example1));
    console.log('');

    const example2 = "2";
    console.log(`Input: "${example2}"`);
    console.log('Output:', solve(example2));

    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This is a classic backtracking problem
 * - The solution explores all possible combinations systematically
 * - Can be visualized as a tree where each level represents a digit
 * - Each branch represents choosing a letter for that digit
 */
